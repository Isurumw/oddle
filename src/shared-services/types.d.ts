declare interface IStorage {
  get(key: string): Promise<string | null>;
  getObject<P>(key: string): Promise<P | null>;
  set(key: string, value: string): Promise<void>;
  remove(key: string): void;
}

declare interface INetworkManager {
  request(
    query: string,
    variables?: object | null,
    method?: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'PATCH',
  ): any;
}

declare type Product = {
  id: string;
  name?: string;
  description?: string;
  rating?: number;
  price?: string;
  brand?: string;
  tagList: string[];
  category?: string;
  productType?: string;
  apiFeaturedImage?: string;
  websiteLink?: string;
  productLink?: string;
};

declare type FavoriteIdObject = {
  id: string;
  liked: boolean;
};

type ProductFetchType = 'productType' | 'brand' | 'category';
