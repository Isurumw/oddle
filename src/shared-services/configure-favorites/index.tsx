import {RestNetworkManager} from 'shared-services';

type Response = {
  success?: boolean;
};

export const configureFavorites = async (
  id: string,
  state: 'add' | 'remove',
): Promise<boolean> => {
  const response = await RestNetworkManager.request<Response>(
    `favourites/${id}`,
    undefined,
    state === 'add' ? 'PATCH' : 'DELETE',
  );
  return response.success ?? false;
};
