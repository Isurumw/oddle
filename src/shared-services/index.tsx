import AsyncStorage from '@react-native-async-storage/async-storage';
import Config from 'react-native-config';

import _StorageService from './storage-service';
import _NetworkManager from './network-services';

const StorageService = new _StorageService(AsyncStorage);

const RestNetworkManager = new _NetworkManager(
  Config.REST_BASEURL,
  fetch,
  Config.REST_AUTH_TOKEN,
  Config.REST_ACCOUNT_NAME,
);

const GraphQLNetworkManager = new _NetworkManager(
  Config.GRAPH_BASEURL,
  fetch,
  Config.GRAPH_AUTH_TOKEN,
);

export {StorageService, RestNetworkManager, GraphQLNetworkManager};
