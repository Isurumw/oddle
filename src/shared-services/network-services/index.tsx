class NetworkManager implements INetworkManager {
  private base: string;
  private fetch: (
    input: RequestInfo,
    init?: RequestInit | undefined,
  ) => Promise<Response>;
  private token?: string;

  public constructor(
    base: string,
    fetch: any,
    token?: string,
    account?: string,
  ) {
    this.fetch = fetch;
    this.token = token;
    this.base = `${base}${account ? `${account}/` : ''}`;
  }

  public async request<T>(
    query: string,
    variables?: object,
    method?: 'POST' | 'GET' | 'PUT' | 'DELETE' | 'PATCH',
  ): Promise<T> {
    try {
      const isGraph = !method;
      const url = `${this.base}${!isGraph ? query : ''}`;

      let headers = {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      };

      let body: {} = {};

      if (isGraph) {
        body = {
          query,
        };
      }

      if (variables && isGraph) {
        body = {
          ...body,
          variables,
        };
      }

      if (this.token) {
        headers = {
          ...headers,
          ...{Authorization: `${isGraph ? 'Bearer ' : ''}${this.token}`},
        };
      }

      const result = await this.fetch(url, {
        method: method ?? 'POST',
        headers,
        body:
          method !== 'GET' ? JSON.stringify(!isGraph ? variables : body) : null,
      });
      return result.json();
    } catch (e: any) {
      throw new Error(e);
    }
  }
}

export default NetworkManager;
