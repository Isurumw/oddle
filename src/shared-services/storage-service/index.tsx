/*
 * The "IStorage" is a simple wrapper around AsyncStorage storage. Use this class to context and
 * retrieve user-data.
 */

class StorageService<T> implements IStorage {
  private storage: any;

  public constructor(storage: T) {
    this.storage = storage;
  }

  public async get(key: string): Promise<string | null> {
    try {
      const value = await this.storage.getItem(key);
      return value || null;
    } catch (err) {
      throw new Error(`Could not receive the ${key} from cache`);
    }
  }

  public async getObject<P>(key: string): Promise<P | null> {
    try {
      const value = await this.storage.getItem(key);
      return value ? JSON.parse(value) : null;
    } catch (err) {
      throw new Error(`Could not receive the ${key} from cache`);
    }
  }

  public async set(key: string, value: string): Promise<void> {
    try {
      await this.storage.setItem(key, value);
    } catch (err) {
      throw new Error(`Could not save the ${key} in cache`);
    }
  }

  public async remove(key: string): Promise<void> {
    try {
      await this.storage.removeItem(key);
    } catch (err) {
      throw new Error(`Could not remove the ${key} from cache`);
    }
  }
}

export default StorageService;
