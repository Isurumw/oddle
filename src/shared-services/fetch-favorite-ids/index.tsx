import {RestNetworkManager} from 'shared-services';

type Response = {
  data: string[];
};

export const fetchFavoriteIds = async (): Promise<string[]> => {
  const response = await RestNetworkManager.request<Response>(
    'favourites',
    undefined,
    'GET',
  );
  return response.data ?? [];
};
