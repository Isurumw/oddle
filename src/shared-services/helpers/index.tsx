import {Linking} from 'react-native';
import {InAppBrowser} from 'react-native-inappbrowser-reborn';
import {StorageService} from 'shared-services';

import {IBM_PLEX_SANS_REGULAR, IBM_PLEX_SANS_BOLD} from 'styles/fonts';
import {PRODUCT_TYPES} from 'utils/constants';

export const fetchFontFamily = (isBold: boolean) => {
  return isBold ? IBM_PLEX_SANS_BOLD : IBM_PLEX_SANS_REGULAR;
};

export const fetchFontWeight = (isBold: boolean) => {
  return isBold ? '700' : '400';
};

export const capitalize = (text: string): string => {
  const word = text.split('_').join(' ');
  const lowerCaseText = word.toLowerCase();
  return lowerCaseText.charAt(0).toUpperCase() + lowerCaseText.slice(1);
};

export const openWeb = async (url?: string) => {
  if (!url) throw new Error('The url is not available');
  if (await InAppBrowser.isAvailable()) {
    InAppBrowser.openAuth(url, url, {
      // iOS Properties
      ephemeralWebSession: false,
      // Android Properties
      showTitle: false,
      enableUrlBarHiding: true,
      enableDefaultShare: false,
    }).then(response => {
      if (response.type === 'success' && response.url)
        Linking.openURL(response.url);
    });
  }
};

export const fetchFavoriteObjects = (ids: string[]) => {
  const objects: FavoriteIdObject[] = ids.map(id => {
    return {id: id, liked: true};
  });

  return objects;
};

export const filterOutFavoriteIds = (objects: FavoriteIdObject[]) => {
  const ids: string[] = objects.map(object => {
    return object.id;
  });

  return ids;
};

export const isFavoriteId = (objects: FavoriteIdObject[], id: string) => {
  const filtered = objects.filter(object => object.id === id && object.liked);
  return filtered.length > 0;
};

export const configureLocalFavorites = (
  objects: FavoriteIdObject[],
  id: string,
  isFavorite?: boolean,
) => {
  const _isFavorite = isFavorite ?? isFavoriteId(objects, id);
  const filtered = objects.filter(object => object.id !== id);

  return filtered.concat([{id: id, liked: !_isFavorite}]);
};

export const filteredField = (
  field: string,
  array: string[],
  state: 'add' | 'remove',
) => {
  const _array = Array.from(array);
  if (state === 'add') _array.push(field);
  else
    _array.splice(
      _array.findIndex(b => b === field),
      1,
    );
  return _array;
};

export const highestOccurrence = (array: string[]) => {
  const _array = Array.from(array);
  return _array
    .sort(
      (a, b) =>
        _array.filter(v => v === a).length - _array.filter(v => v === b).length,
    )
    .pop();
};
