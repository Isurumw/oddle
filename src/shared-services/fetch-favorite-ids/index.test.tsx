import {RestNetworkManager} from 'shared-services';
import {fetchFavoriteIds} from './';

jest.mock('shared-services/network-services');
const mockedRestService =
  RestNetworkManager as unknown as jest.Mocked<INetworkManager>;

const mockResponse = {
  data: ['cl4cevbjy3liu0cpnim92e5eo'],
};

describe('check the behavior of the network service:', () => {
  it('fetched product ids successfully', async () => {
    mockedRestService.request.mockImplementation(() =>
      Promise.resolve(mockResponse),
    );
    const response = await fetchFavoriteIds();

    expect(mockedRestService.request).toHaveBeenCalledTimes(1);
    expect(mockedRestService.request).toMatchSnapshot();
    expect(response).toEqual(mockResponse.data);
  });

  it('When API throws an error', async () => {
    mockedRestService.request.mockImplementation(() =>
      Promise.reject(new Error('Some API Error')),
    );

    expect(fetchFavoriteIds()).rejects.toMatchSnapshot();
  });
});
