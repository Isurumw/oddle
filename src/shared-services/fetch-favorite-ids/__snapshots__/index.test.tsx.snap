// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check the behavior of the network service: When API throws an error 1`] = `[Error: Some API Error]`;

exports[`check the behavior of the network service: fetched product ids successfully 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "favourites",
      undefined,
      "GET",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_U": 0,
        "_V": 1,
        "_W": Object {
          "data": Array [
            "cl4cevbjy3liu0cpnim92e5eo",
          ],
        },
        "_X": null,
      },
    },
  ],
}
`;
